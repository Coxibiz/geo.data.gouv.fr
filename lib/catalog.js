import moment from 'moment'
import { get, filter, sortBy } from 'lodash'

export const isObsolete = catalog => {
  const revisionDate = get(catalog, 'metrics.mostRecentRevisionDate')

  return revisionDate && moment().subtract(6, 'months').isAfter(revisionDate)
}

export const findCandidates = (catalogs, blacklist = []) => filter(
  catalogs,
  catalog =>
    !blacklist.includes(catalog.id) &&
    get(catalog, 'metrics.datasets.partitions.openness.yes', 0) >= 1 &&
    get(catalog, 'metrics.datasets.partitions.download.yes', 0) >= 1
)

export const computeScore = catalog => {
  const total = get(catalog.metrics, 'datasets.totalCount', 0)

  if (!total) {
    return 0
  }

  const percentOpen = get(catalog.metrics, 'datasets.partitions.openness.yes', 0) / total * 100
  const percentDownloadable = get(catalog.metrics, 'datasets.partitions.download.yes', 0) / total * 100

  const now = new Date()
  const sixMonthsAgo = moment(now).subtract(6, 'months')
  const oneMonthAgo = moment(now).subtract(1, 'months')

  // Compute freshness
  let freshness = 1
  if (catalog.metrics.mostRecentRevisionDate) {
    if (oneMonthAgo.isBefore(catalog.metrics.mostRecentRevisionDate)) {
      freshness = 100
    } else if (sixMonthsAgo.isBefore(catalog.metrics.mostRecentRevisionDate)) {
      freshness = 10
    }
  }

  // Compute downloadability
  const downloadability = Math.pow(Math.max(Math.min(percentDownloadable, 100), 1), 2)

  // Compute openness
  const openness = Math.max(Math.min(percentOpen, 100), 1)

  return freshness * downloadability * openness
}

export const sortByScore = catalogs => sortBy(catalogs, catalog => -computeScore(catalog))
