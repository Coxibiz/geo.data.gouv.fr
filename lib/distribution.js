import getConfig from 'next/config'

const {publicRuntimeConfig: {
  GEODATA_API_URL
}} = getConfig()

export function generateDistributionInfo(distribution) {
  let link
  let layerName

  switch (distribution.type) {
    case 'file-package': {
      const idx = distribution.layer.lastIndexOf('/')

      layerName = idx === -1 ?
        distribution.layer :
        distribution.layer.substring(idx + 1, distribution.layer.length)

      link = `${GEODATA_API_URL}/file-packages/${distribution.hashedLocation}/${layerName}/download`
      break
    }

    case 'wfs-featureType':
      link = `${GEODATA_API_URL}/services/${distribution.service}/feature-types/${distribution.typeName}/download`
      break

    default:
      break
  }

  return {
    link,
    layerName: layerName || distribution.typeName
  }
}

export function generateDistributionId(distribution) {
  switch (distribution.type) {
    case 'file-package':
      return `file-package:${distribution.hashedLocation}/${distribution.layer}`

    case 'wfs-featureType':
      return `wfs:${distribution.service}/${distribution.typeName}`

    default:
      return null
  }
}
