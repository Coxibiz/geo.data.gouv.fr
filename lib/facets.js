export const facetTypes = [
  'availability',
  'dgvPublication',
  'distributionFormat',
  'keyword',
  'metadataType',
  'opendata',
  'organization',
  'representationType',
  'type',
  'catalog',
  'license'
]
