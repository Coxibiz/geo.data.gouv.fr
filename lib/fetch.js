import 'isomorphic-unfetch'

import { DATAGOUV_API_URL, DATAGOUV_API_KEY, PUBLICATION_BASE_URL } from '@env'

class HttpError extends Error {
  constructor(message, code) {
    super(message)

    this.code = code
  }
}

export function _fetch(url, method, data) {
  const options = {
    headers: {
      'Accept': 'application/json'
    },
    mode: 'cors',
    method: method || 'GET'
  }

  if (url.includes(PUBLICATION_BASE_URL + '/api')) {
    options.credentials = 'include'
  }

  if (url.startsWith(DATAGOUV_API_URL)) {
    if (DATAGOUV_API_KEY) {
      options.headers['X-API-KEY'] = DATAGOUV_API_KEY
      options.mode = undefined
    } else {
      options.credentials = 'include'
    }
  }

  if (data) {
    options.headers['Content-Type'] = 'application/json'
    options.body = JSON.stringify(data)
  }

  return fetch(url, options).then(response => {
    if (response.status === 502) throw new HttpError('Bad Gateway', 502)
    if (response.status === 500) throw new HttpError('Internal Server Error', 500)
    if (response.status === 401) throw new HttpError('Unauthorized', 401)
    if (response.status === 404) throw new HttpError('Not found', 404)
    if (response.status === 202) return
    if (response.status === 204) return
    return response.json()
  })
}

export function _get(url) {
  return _fetch(url)
}

export function _put(url, data) {
  return _fetch(url, 'PUT', data)
}

export function _post(url, data) {
  return _fetch(url, 'POST', data)
}

export function _delete(url) {
  return _fetch(url, 'DELETE')
}
