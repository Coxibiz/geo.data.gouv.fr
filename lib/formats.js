export default [
  {
    label: 'GeoJSON',
    format: 'GeoJSON',
    projection: 'WGS84'
  },
  {
    label: 'SHP/L93',
    format: 'SHP',
    projection: 'Lambert93'
  },
  {
    label: 'SHP/W84',
    format: 'SHP',
    projection: 'WGS84'
  },
  {
    label: 'KML',
    format: 'KML',
    projection: 'WGS84'
  },
  {
    label: 'CSV',
    format: 'CSV',
    projection: 'WGS84'
  }
]
