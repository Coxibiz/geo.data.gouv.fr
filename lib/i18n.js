import i18n from 'i18next'
import LanguageDetector from 'i18next-browser-languagedetector'
import XHR from 'i18next-xhr-backend'
import moment from 'moment'

const dev = process.env.NODE_ENV !== 'production'

const languages = [
  'fr',
  'en'
]

if (!i18n.isInitialized) {
  i18n
    .use(XHR)
    .use(LanguageDetector)
    .init({
      fallbackLng: 'fr',
      load: 'languageOnly',
      lowerCaseLng: true,

      ns: ['common'],
      defaultNS: 'common',
      whitelist: [...languages],

      saveMissing: dev,

      interpolation: {
        escapeValue: false,
        formatSeparator: ',',
        format: (value, format, lng) => {
          if (format === 'uppercase') {
            return value.toUpperCase()
          }

          return value
        }
      },

      detection: {
        caches: ['cookie'],
        order: []
      }
    })

  i18n.on('languageChanged', lang => {
    moment.locale(lang)
  })
}

export default i18n
