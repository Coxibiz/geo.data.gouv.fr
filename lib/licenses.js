export default {
  'fr-lo': {
    name: 'Licence Ouverte',
    link: 'https://www.etalab.gouv.fr/licence-ouverte-open-licence'
  },

  'fr-loa': {
    name: 'Licence Ouverte Administrations',
    link: 'https://www.etalab.gouv.fr/nouvelle-licence-pour-la-reutilisation-des-informations-publiques-elements-de-clarification'
  },

  'odc-odbl': {
    name: 'Open Database License (ODbL 1.0)',
    link: 'http://vvlibri.org/fr/licence/odbl-10/legalcode/unofficial'
  }
}
