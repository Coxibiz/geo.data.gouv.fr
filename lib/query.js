export const ignoredFilters = [
  'q',
  'page',
  'offset',
  'limit'
]

export function getFilters(query) {
  const filters = []

  Object
    .entries(query)
    .filter(([name]) => !ignoredFilters.includes(name))
    .forEach(([name, value]) => {
      if (Array.isArray(value)) {
        value.forEach(v => filters.push({
          name,
          value: v
        }))
      } else {
        filters.push({name, value})
      }
    })

  return filters.reduce((acc, filter) => {
    (acc[filter.name] = acc[filter.name] || []).push(filter.value)
    return acc
  }, {})
}
