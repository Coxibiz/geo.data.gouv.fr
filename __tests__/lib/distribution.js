import {generateDistributionInfo, generateDistributionId} from '../../lib/distribution'

describe('generateDistributionInfo', () => {
  it('should generate distribution link and layerName for a wfs distribution', () => {
    const distribution = {
      _id: '5ae9b6c7708c195caaf9cd79',
      type: 'wfs-featureType',
      service: '5a2e9c73f43c6a15cb04f897',
      typeName: 'hdf_draaf:l_part_sau_2015_cantons_s_r32',
      available: true
    }

    expect(generateDistributionInfo(distribution)).toMatchSnapshot()
  })

  it('should generate distribution link and layerName for a file-package distribution', () => {
    const distribution = {
      _id: '5ae7bcf390e0ad57ec364c48',
      type: 'file-package',
      name: 'hydrant',
      location: 'http://ids.e-aveyron.fr/media/easysdi/images/17910f42-0c5b-fee4-1503-69a707b72874_pei%20%2811%29.zip',
      hashedLocation: '8a704988cc8000c54c9b36761d681674c7ac4b24',
      available: true,
      layer: 'peiPoint.shp'
    }

    expect(generateDistributionInfo(distribution)).toMatchSnapshot()
  })

  it('should generate undefined properties if the distribution type is unknown', () => {
    const distribution = {
      type: 'something'
    }

    expect(generateDistributionInfo(distribution)).toMatchSnapshot()
  })

  it('should assign the typeName to layerName is defined', () => {
    const distribution = {
      type: 'something',
      typeName: 'foo'
    }

    expect(generateDistributionInfo(distribution)).toMatchSnapshot()
  })
})

describe('generateDistributionId', () => {
  it('should generate an identifier for file-package distributions', () => {
    const distribution = {
      _id: '5ae9b6c7708c195caaf9cd79',
      type: 'wfs-featureType',
      service: '5a2e9c73f43c6a15cb04f897',
      typeName: 'hdf_draaf:l_part_sau_2015_cantons_s_r32',
      available: true
    }

    expect(generateDistributionId(distribution)).toBe(
      'wfs:5a2e9c73f43c6a15cb04f897/hdf_draaf:l_part_sau_2015_cantons_s_r32'
    )
  })

  it('should generate an identifier for wfs distributions', () => {
    const distribution = {
      _id: '5ae7bcf390e0ad57ec364c48',
      type: 'file-package',
      name: 'hydrant',
      location: 'http://ids.e-aveyron.fr/media/easysdi/images/17910f42-0c5b-fee4-1503-69a707b72874_pei%20%2811%29.zip',
      hashedLocation: '8a704988cc8000c54c9b36761d681674c7ac4b24',
      available: true,
      layer: 'peiPoint.shp'
    }

    expect(generateDistributionId(distribution)).toBe(
      'file-package:8a704988cc8000c54c9b36761d681674c7ac4b24/peiPoint.shp'
    )
  })

  it('should not generate anything if the distribution type is unknown', () => {
    const distribution = {
      type: 'foo'
    }

    expect(generateDistributionId(distribution)).toBe(null)
  })
})
