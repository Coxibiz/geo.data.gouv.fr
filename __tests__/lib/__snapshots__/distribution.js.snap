// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateDistributionInfo should assign the typeName to layerName is defined 1`] = `
Object {
  "layerName": "foo",
  "link": undefined,
}
`;

exports[`generateDistributionInfo should generate distribution link and layerName for a file-package distribution 1`] = `
Object {
  "layerName": "peiPoint.shp",
  "link": "undefined/file-packages/8a704988cc8000c54c9b36761d681674c7ac4b24/peiPoint.shp/download",
}
`;

exports[`generateDistributionInfo should generate distribution link and layerName for a wfs distribution 1`] = `
Object {
  "layerName": "hdf_draaf:l_part_sau_2015_cantons_s_r32",
  "link": "undefined/services/5a2e9c73f43c6a15cb04f897/feature-types/hdf_draaf:l_part_sau_2015_cantons_s_r32/download",
}
`;

exports[`generateDistributionInfo should generate undefined properties if the distribution type is unknown 1`] = `
Object {
  "layerName": undefined,
  "link": undefined,
}
`;
